Re-write DateTimeType trait with DateTimeInput

sffc opened the pull request:
Fixes #355

nordzilla commented:
@sffc this does have some overlap with the design document that I'm writing for timezones. 

In the deep dive, we had discussed separating the traits out into composable functionality:

> Zibi: "I suggest we don't do one trait (interface). Ex: a zoned DateTime could have a Date trait, a Time trait, and Zone trait. This flexibility is also what modern datetime libraries do as well."
>
> Shane: "I would like to be more granular/general than just a tuple of optional Time, Date, and Zone. We might have other fields that we want to support."

I had a trait dependency tree like this in mind, which you could potentially implement here (minus the timezones):

```
trait ZonedDateTimeType
│
├──── trait DateTimeType
│     │
│     ├──── trait FromStr
│     │     │
│     │     └──── trait Sized
│     │
│     ├──── trait DateFormat
│     │
│     └──── trait TimeFormat
│
└──── trait ZoneFormat
```

For the purpose of this PR, you would effectively have:

```rust
pub trait DateFormat {
  // Functionality required for date formatting
}

pub trait TimeFormat {
  // Functionality required for time formatting
}

pub trait DateTimeType: DateFormat + TimeFormat + FromStr {}

impl<T> DateTimeType for T
  where T: TimeFormat + DateFormat + FromStr {}
```

Thoughts?

sffc commented:
That's a really good point.  I'll go ahead and split up the trait.

sffc commented:
This is ready for review.  I wrote the new DateTime input trait consistent with previous discussions and implemented it on MockDateTime.  I removed the old trait and rewired the formatting code as needed.

sffc commented:
2021-02-04: @zbraniecki and @nordzilla will write down ideas on how to structure the date time input traits.

nordzilla reviewed (approved):
This looks good to me, plus a few thoughts, questions, and non-blocking suggestions.

zbraniecki commented on components/datetime/src/date.rs:
```diff
@@ -28,152 +30,69 @@ impl From<std::num::ParseIntError> for DateTimeError {
-            minute,
-            second,
+            data,
+            // TODO
```
May I take a moment of your time to preach the API design of Rust to you? :) `todo!("sffc, #451")` seems like a match for that.

We could even derive from that to enforce the culture `icu4x_todo!(reported, issue);`

nordzilla replied:
@zbraniecki 

Love this idea.

Sure, you can always get the reporter from the blame, but I always find it inconvenient after arbitrary "ran cargo format"-like patches, when it's really easy to just add that in, especially if you happen to be reading a snippet of code without blame context.

sffc replied:
`todo!` looks cool, but it panics.  In most cases I just want to leave a comment.

At Google we use the `// TODO(ticketnumber)` approach.  I'll do that here.

sffc replied:
I added a ticket number to all the TODOs, and replaced `unimplemented!` with `todo!` where appropriate.

zbraniecki commented on components/datetime/src/format.rs:
```diff
@@ -83,112 +88,199 @@ where
+            .take_payload()
+            .unwrap();
+        let pattern = crate::pattern::Pattern::from_bytes("MMM").unwrap();
+        let date_time = MockDateTime::try_new(2020, 8, 1, 12, 34, 28).unwrap();
```
I'm strongly against overuse of the builder pattern here. Builder pattern is the right choice if the state of the struct makes sense in between various calls. In this case the `MockDateTime` makes no sense with half of the field initialized, and using a builder pattern for that is imho a bad API design model (which I've seen way too often).

Instead, we could use a constructor options struct:

```rust
MockDateTime::try_new(MockDateTimeOptions {
    year: 2020,
    month: 8,
    day: 1,
    hour: 12,
    minute: 34,
    second: 28
});
```

which is analogous to builder pattern in the expressiveness but doesn't bring the problem of half-initialized states.

In this particular case tho, I think that a long list of arguments is fine, since it's a mock and we don't intend it to be the final user facing pattern in the long run.

nordzilla replied:
I admit to being guilty of "overusing" the builder pattern for only the sake of readability. I just find it to be incredibly readable and expressive. 

But I don't disagree with what you just said.

zbraniecki replied:
The issue with the builder pattern mid-state is two fold:

1) You can, and by the law of nature since you can you will, end up with a half-initialized structs. You can of course end up with `.build()` method that fails if the struct is not ready, but that means you moved architectural error reporting from compile time to run time at a loss.
2) When you change the list of fields required, you have no way to inform your users that they now need to call one more method on the builder pattern, except of, again, failing at `.build()` method, which is runtime.

nordzilla replied:
All valid points. No arguments from me.

sffc replied:
I'm not interested in rewriting MockDateTime in this PR.  I think it should be thrown out in the dumpster because it makes several really bad assumptions, not to mention the fact that it only works in proleptic Gregorian.  When we come up with a real datetime struct, we can engage these design discussions.

sffc commented on components/datetime/src/format.rs:
```diff
@@ -83,112 +88,199 @@ where
+                        .month()
+                        .ok_or(Error::MissingInputField)?
+                        .number as usize
+                        - 1,
```
0-indexed versus 1-indexed months are a nightmare.  Here is what I'm currently trying to enforce:

- 0-indexed:
    - Locale data struct
    - MockDateTime
- 1-indexed:
    - ISO-8601 string
    - DateInput trait

When I add or subtract 1, it's when moving between the two models.

sffc replied:
I added additional documentation explaining when 0-based is used and when 1-based is used.

sffc commented on components/datetime/src/date.rs:
```diff
@@ -267,9 +186,65 @@ macro_rules! dt_unit {
+    Millisecond(u16),
+    Microsecond(u32),
+    Nanosecond(u32),
+}
```
Yes, it should.  I don't want to scope-creep this PR though.  Can you open a follow-up issue for this?

sffc replied:
https://github.com/unicode-org/icu4x/issues/485

zbraniecki commented on components/datetime/src/date.rs:
```diff
@@ -267,9 +186,65 @@ macro_rules! dt_unit {
+    pub const THU: WeekDay = WeekDay(4);
+    pub const FRI: WeekDay = WeekDay(5);
+    pub const SAT: WeekDay = WeekDay(6);
+}
```
why mod with consts instead of `enum`?

sffc commented on components/datetime/src/format.rs:
```diff
@@ -40,35 +42,38 @@ use writeable::Writeable;
 
 // Temporary formatting number with length.
-fn format_number<W>(result: &mut W, num: usize, length: FieldLength) -> Result<(), std::fmt::Error>
+fn format_number<W>(result: &mut W, num: isize, length: &FieldLength) -> Result<(), std::fmt::Error>
```
Not sure. Reverted.

sffc commented:
I responded to all review feedback.  I also added more docs.  PTAL.

nordzilla reviewed (approved):
Thanks @sffc!

Sorry, I didn't mean to leave this hanging. I've been out the past two days. 

This looks good to me.

zbraniecki reviewed (dismissed):
lgtm!

sffc commented on components/datetime/src/arithmetic.rs:
```diff
@@ -0,0 +1,91 @@
+
+/// Temporary simplified function to get the day of the week
+/// month and day are both zero-indexed.
+pub fn iso_date_to_weekday(year: i32, month: usize, day: usize) -> IsoWeekday {
```
Good point.  Since I removed the `< 12` requirement on `month`, there are multiple places where we could trigger an out-of-range error.  This is one; another is in `get_symbol_for_month`.

For now, I'll add `debug_assert!` in these places, but we don't actually enforce this anywhere anymore.  We should follow up with this in https://github.com/unicode-org/icu4x/issues/493.
